src/ORM/Behavior/TimestampBehavior.php

```diff
@@ -130,19 +130,23 @@ public function implementedEvents()
      * If an explicit date time is passed, the config option `refreshTimestamp` is
      * automatically set to false.
      *
-     * @param \DateTime|null $ts Timestamp
+     * @param \DateTimeInterface|null $ts Timestamp
      * @param bool $refreshTimestamp If true timestamp is refreshed.
-     * @return \DateTime
+     * @return \DateTimeInterface
      */
-    public function timestamp(DateTime $ts = null, $refreshTimestamp = false)
+    public function timestamp(DateTimeInterface $ts = null, $refreshTimestamp = false)
     {
+        /** @var \Cake\Database\Type\DateTimeType $type */
+        $type = Type::build('datetime');
```

Comment:
It is better than now, but if the `modified` field is declared as `timestamp`, and the `Type` settings are different between `'datetime'` and `'timestamp'`, I think it doesn't work as expected. 

See: https://github.com/cakephp/app/blob/94f120be6498f05654dfa83f6ea79302fa667661/config/bootstrap.php#L185-L188

Reply:
it slipped my mind..

as default of [Type::$_types](https://github.com/cakephp/cakephp/blob/3.5.10/src/Database/Type.php#L34-L51), `datetime` and `timestamp` both are bound to `DateTimeType`.
(and DateTimeType::$_className value is shared as statically.)

i should consider there are difference  between `created` and `modified` column types ?

Reply:
Hmm. It seems impossible to detect correct types in this method, as it doesn't know what field uses the timestamp. I think it can be done in the `_updateField()` method instead, by using `$this->_table->getSchema()`.

Reply:
We won't be able to get the class correct 100% of the time, but it is more *correct* now than it was in the past.

Reply:
It is true that should plan for a change that actually uses the right type

Reply:
How would we get the correct types based on the arguments to this method?

Reply:
can i add `field` option to `timestamp()` and `_updateField()` passes field name ?

--------

```php
    /**
     * @param string  $field Field name to set timestamp
     */
public function timestamp(DateTimeInterface $ts = null, $refreshTimestamp = false, $field = 'created')
    {
        $type = Type::build('datetime');
        $class = $type->getDateTimeClassName();
```
and
```php
    protected function _updateField($entity, $field, $refreshTimestamp)
    {
        if ($entity->isDirty($field)) {
            return;
        }
        $entity->set($field, $this->timestamp(null, $refreshTimestamp, $field));
    }
}
```

Reply:
if we should use the right type, maybe it would be better to to keep `$_ts` each fields...
because when `$ts` is null && !$refreshTimestamp, `timestamp()` returns cached instance.

Reply:
I think you could revert timestamp() and change `_updateField()` method like the following: 
```
protected function _updateField($entity, $field, $refreshTimestamp)
{
    if ($entity->isDirty($field)) {
        return;
    }

    $ts = $this->timestamp(null, $refreshTimestamp);

    $columnType = $this->getTable()->getScheme()->getColumnType($field);
    $type = Type::build($columnType);
    $class = $type->getDateTimeClassName();

    // Create a new proper DateTimeInterface instance here

    $entity->set($field, $ts);
}
```

Reply:
thanks, i'll  do it!